include/boost/beast/_experimental/test/impl/stream.hpp

```diff
@@ -44,7 +39,7 @@ class basic_stream<Executor>::read_op : public detail::stream_read_op_base
 #if defined(BOOST_ASIO_NO_TS_EXECUTORS)
         net::any_io_executor wg2_;
 #else // defined(BOOST_ASIO_NO_TS_EXECUTORS)
-        net::executor_work_guard<ex2_type> wg2_;
+        net::executor_work_guard<net::any_io_executor> wg2_;
```

Comment:
What is this change for?

Reply:
There was a compile time error in this line:
https://github.com/boostorg/beast/blob/35d533299c129d622af774f96b820e8f18df068f/include/boost/beast/_experimental/test/impl/stream.hpp#L67

The error occurres when the associated executor of the handler is different from the wg2 executor. This only happens during the instantiation of the [is_async_read_stream](https://github.com/boostorg/beast/blob/35d533299c129d622af774f96b820e8f18df068f/include/boost/beast/core/stream_traits.hpp#L337) trait. I believe this is due to the fact that detail::ReadHandler has no associated executor and uses s->exec in that line.

You can find the exact error here:
https://github.com/boostorg/beast/actions/runs/5925847678/job/16066139656#step:9:1423
```BASH
clang-linux.compile.c++ ../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden/accept.o
In file included from test/beast/websocket/accept.cpp:13:
In file included from ../../boost/beast/_experimental/test/stream.hpp:617:
../../boost/beast/_experimental/test/impl/stream.hpp:67:15: error: no matching constructor for initialization of 'net::executor_work_guard<ex2_type>' (aka 'executor_work_guard<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>')
            , wg2_(net::get_associated_executor(
              ^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../boost/beast/_experimental/test/impl/stream.hpp:144:11: note: in instantiation of function template specialization 'boost::beast::test::basic_stream<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>::read_op<boost::beast::detail::StreamHandler, boost::beast::detail::BufferSequence<boost::asio::mutable_buffer>>::lambda::lambda<boost::beast::detail::StreamHandler>' requested here
        : fn_(std::forward<Handler_>(h), s, b)
          ^
../../boost/beast/_experimental/test/impl/stream.hpp:192:17: note: in instantiation of function template specialization 'boost::beast::test::basic_stream<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>::read_op<boost::beast::detail::StreamHandler, boost::beast::detail::BufferSequence<boost::asio::mutable_buffer>>::read_op<boost::beast::detail::StreamHandler>' requested here
            new read_op<
                ^
../../boost/asio/async_result.hpp:482:5: note: in instantiation of function template specialization 'boost::beast::test::basic_stream<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>::run_read_op::operator()<boost::beast::detail::StreamHandler, boost::beast::detail::BufferSequence<boost::asio::mutable_buffer>>' requested here
    BOOST_ASIO_MOVE_CAST(Initiation)(initiation)(
    ^
../../boost/asio/detail/config.hpp:135:37: note: expanded from macro 'BOOST_ASIO_MOVE_CAST'
# define BOOST_ASIO_MOVE_CAST(type) static_cast<type&&>
                                    ^
../../boost/asio/async_result.hpp:896:21: note: in instantiation of function template specialization 'boost::asio::detail::completion_handler_async_result<boost::beast::detail::StreamHandler, void (boost::system::error_code, unsigned long)>::initiate<boost::beast::test::basic_stream<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>::run_read_op, boost::beast::detail::StreamHandler, boost::shared_ptr<boost::beast::test::detail::stream_state> &, const boost::beast::detail::BufferSequence<boost::asio::mutable_buffer> &>' requested here
    Signatures...>::initiate(BOOST_ASIO_MOVE_CAST(Initiation)(initiation),
                    ^
../../boost/beast/_experimental/test/impl/stream.hpp:343:17: note: in instantiation of function template specialization 'boost::asio::async_initiate<boost::beast::detail::StreamHandler, void (boost::system::error_code, unsigned long), boost::beast::test::basic_stream<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>::run_read_op, boost::shared_ptr<boost::beast::test::detail::stream_state> &, const boost::beast::detail::BufferSequence<boost::asio::mutable_buffer> &>' requested here
    return net::async_initiate<
                ^
../../boost/asio/executor_work_guard.hpp:97:12: note: candidate constructor not viable: no known conversion from 'const boost::asio::detail::associated_executor_impl<boost::beast::detail::StreamHandler, boost::asio::any_io_executor>::type' (aka 'const boost::asio::any_io_executor') to 'const boost::asio::executor_work_guard<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>::executor_type' (aka 'const boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>') for 1st argument
  explicit executor_work_guard(const executor_type& e) BOOST_ASIO_NOEXCEPT
           ^
../../boost/asio/executor_work_guard.hpp:104:3: note: candidate constructor not viable: no known conversion from 'const boost::asio::detail::associated_executor_impl<boost::beast::detail::StreamHandler, boost::asio::any_io_executor>::type' (aka 'const boost::asio::any_io_executor') to 'const boost::asio::executor_work_guard<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>' for 1st argument
  executor_work_guard(const executor_work_guard& other) BOOST_ASIO_NOEXCEPT
  ^
../../boost/asio/executor_work_guard.hpp:113:3: note: candidate constructor not viable: no known conversion from 'const boost::asio::detail::associated_executor_impl<boost::beast::detail::StreamHandler, boost::asio::any_io_executor>::type' (aka 'const boost::asio::any_io_executor') to 'boost::asio::executor_work_guard<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>' for 1st argument
  executor_work_guard(executor_work_guard&& other) BOOST_ASIO_NOEXCEPT
  ^
1 error generated.

    "clang++-12"   -std=c++14 -fvisibility-inlines-hidden -fPIC -m64 -pthread -O0 -fno-inline -Wall -g -fvisibility=hidden --target=x86_64-pc-linux  -DBOOST_ALL_NO_LIB=1 -DBOOST_ASIO_DISABLE_BOOST_ARRAY=1 -DBOOST_ASIO_DISABLE_BOOST_BIND=1 -DBOOST_ASIO_DISABLE_BOOST_DATE_TIME=1 -DBOOST_ASIO_DISABLE_BOOST_REGEX=1 -DBOOST_ASIO_NO_DEPRECATED=1 -DBOOST_ASIO_SEPARATE_COMPILATION -DBOOST_ATOMIC_DYN_LINK=1 -DBOOST_BEAST_ALLOW_DEPRECATED -DBOOST_BEAST_SEPARATE_COMPILATION -DBOOST_BEAST_TESTS -DBOOST_CONTEXT_DYN_LINK=1 -DBOOST_COROUTINES_NO_DEPRECATION_WARNING=1 -DBOOST_FILESYSTEM_DYN_LINK=1 -D_GNU_SOURCE=1 -D_XOPEN_SOURCE=600   -I"." -I"../.." -I"test/extern" -I"test/extras/include"  -c -o "../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden/accept.o" "test/beast/websocket/accept.cpp"

...failed clang-linux.compile.c++ ../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden/accept.o...
...skipped <p../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden>accept for lack of <p../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden>accept.o...
...skipped <p../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden>accept.run for lack of <p../../bin.v2/libs/beast/test/beast/websocket/accept.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden>accept...
clang-linux.link ../../bin.v2/libs/beast/test/beast/websocket/any_completion_handler.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden/any_completion_handler
testing.capture-output ../../bin.v2/libs/beast/test/beast/websocket/any_completion_handler.test/clang-linux-12/debug/cxxstd-14-iso/threading-multi/visibility-hidden/any_completion_handler.run

```